sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -185,6 +185,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
       // PropagateEmptyRelation can change the nullability of an attribute from nullable to
       // non-nullable when an empty relation child of a Union is removed
       UpdateAttributeNullability) ::
+    Batch("Rewrite with Correlated Expressions", Once,
+      RewriteAsOfJoin) ::
```

Comment:
can we put it in the `Finish Analysis` batch?

Reply:
Sounds good. Let me try and move it there.